### PR TITLE
ORC-1385: [C++] Support schema evolution of numeric types

### PR DESCRIPTION
**Pull request by ffacs**
### What changes were proposed in this pull request?
support schema evolution converting between types within {boolean, tinyint, smallint, int, bigint, float, double}

### Why are the changes needed?
To support shecma evolution in c++

### How was this patch tested?
UT passed

### CODE REVIEW

File: c++/src/ConvertColumnReader.cc [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+  static inline void setNull(ColumnVectorBatch& dstBatch, uint64_t idx) {
+    dstBatch.notNull.data()[idx] = 0;
+    dstBatch.hasNulls = true;
+  }
+
+  // return false if overflow
+  template <typename ReadType>
+  static bool downCastToInteger(ReadType& dstValue, int64_t inputLong) {
+    dstValue = static_cast<ReadType>(inputLong);
+    if (std::is_same<ReadType, int64_t>::value) {
```
Comment (wgtmac) [inline c++/src/ConvertColumnReader.cc:69]:
```suggestion
    if constexpr (std::is_same<ReadType, int64_t>::value) {
```

File: c++/src/ConvertColumnReader.cc [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+      return true;
+    }
+    if (static_cast<int64_t>(dstValue) != inputLong) {
+      return false;
+    }
+    return true;
+  }
+
+  // set null if overflow
```
Comment (wgtmac) [inline c++/src/ConvertColumnReader.cc:78]: Should we add a mode to support nullify or throw when overflow is detected? Different engines may expect different behaviors.

**Review by wgtmac — CHANGES_REQUESTED** [review]

File: c++/include/orc/Vector.hh [hunk 3]
```diff
@@ -53,6 +53,10 @@ namespace orc {
     // whether the vector batch is encoded
     bool isEncoded;
 
+    // whether the vector is not a numeric vector
+    // or it's created with the option useTightNumericVector
+    bool isTight = true;
```
Comment (wgtmac) [inline c++/include/orc/Vector.hh:58]: Is it possible not to add this? The reader can get this information by the provided `ReaderOption`

File: c++/src/ColumnReader.cc [hunk 4]
```diff
@@ -1699,7 +1712,21 @@ namespace orc {
...
+                                            bool useTightNumericVector,
+                                            bool throwOnSchemaEvolutionOverflow,
+                                            bool convertToReadType) {
+    if (convertToReadType && stripe.getSchemaEvolution()) {
+      if (stripe.getSchemaEvolution()->needConvert(type)) {
+        if (!useTightNumericVector) {
+          throw SchemaEvolutionError(
+              "SchemaEvolution only support tight vector, please create ColumnVectorBatch with "
+              "option "
+              "useTightNumericVector");
```
Comment (wgtmac) [inline c++/src/ColumnReader.cc:1724]:
```suggestion
              "SchemaEvolution only support tight vector, please create ColumnVectorBatch with "
              "option useTightNumericVector");
```

File: c++/src/ColumnReader.cc [hunk 5]
```diff
@@ -1699,7 +1712,21 @@ namespace orc {
    * Create a reader for the given stripe.
    */
   std::unique_ptr<ColumnReader> buildReader(const Type& type, StripeStreams& stripe,
-                                            bool useTightNumericVector) {
+                                            bool useTightNumericVector,
+                                            bool throwOnSchemaEvolutionOverflow,
+                                            bool convertToReadType) {
+    if (convertToReadType && stripe.getSchemaEvolution()) {
```
Comment (wgtmac) [inline c++/src/ColumnReader.cc:1718]: Can we merge these if statements to a single one?

File: c++/src/ConvertColumnReader.cc [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+
+namespace orc {
+
+  // Assume that we are using tight numeric vector batch
+  using BooleanVectorBatch = ByteVectorBatch;
+
+  ConvertColumnReader::ConvertColumnReader(const Type& _readType, const Type& fileType,
+                                           StripeStreams& stripe, bool _throwOnOverflow)
+      : ColumnReader(_readType, stripe), readType(_readType), throwOnOverflow(_throwOnOverflow) {
+    reader = buildReader(fileType, stripe, /*useTightNumericVector=*/true,
```
Comment (wgtmac) [inline c++/src/ConvertColumnReader.cc:29]: What about add an input parameter `useTightNumericVector` to the function and throw if it is false? In this way, we don't have to check it any more like line 36 does.

File: c++/src/ConvertColumnReader.cc [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+      throw SchemaEvolutionError(
+          "SchemaEvolution only support tight vector, please create ColumnVectorBatch with option "
+          "useTightNumericVector");
+    }
+    reader->next(*data, numValues, notNull);
+    rowBatch.resize(data->capacity);
+    rowBatch.numElements = data->numElements;
+    rowBatch.hasNulls = data->hasNulls;
+    if (!rowBatch.hasNulls) {
+      memset(rowBatch.notNull.data(), 1, data->notNull.size());
```
Comment (wgtmac) [inline c++/src/ConvertColumnReader.cc:46]: I guess this is not required

File: c++/src/ConvertColumnReader.cc [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+    constexpr bool isReadTypeFloatingPoint(std::is_floating_point<ReadType>::value);
+    int64_t longValue = static_cast<int64_t>(srcValue);
+    if (isFileTypeFloatingPoint) {
+      if (isReadTypeFloatingPoint) {
+        destValue = static_cast<ReadType>(srcValue);
+      } else {
+        if (!canFitInLong(static_cast<double>(srcValue)) ||
+            !downCastToInteger(destValue, longValue)) {
+          handleOverflow<FileType, ReadType>(destBatch, idx, shouldThrow);
+          return;
```
Comment (wgtmac) [inline c++/src/ConvertColumnReader.cc:108]: This line is redundant.

File: c++/src/Reader.cc [hunk 9]
```diff
@@ -516,6 +518,13 @@ namespace orc {
     return forcedScaleOnHive11Decimal;
   }
 
+  void RowReaderImpl::getReadColumns(const Type* readType, std::set<uint64_t>& readColumns) const {
```
Comment (wgtmac) [inline c++/src/Reader.cc:521]: nit: `getReadColumns` can be a non-member function.

File: c++/src/Reader.cc [hunk 10]
```diff
@@ -1205,8 +1215,32 @@ namespace orc {
   }
 
   std::unique_ptr<ColumnVectorBatch> RowReaderImpl::createRowBatch(uint64_t capacity) const {
-    return getSelectedType().createRowBatch(capacity, *contents->pool, enableEncodedBlock,
-                                            useTightNumericVector);
+    if (schemaEvolution.getReadType() && selectedSchema.get() == nullptr) {
```
Comment (wgtmac) [inline c++/src/Reader.cc:1218]: nit: use `RowReaderImpl::getSelectedType()` to get selectedSchema so we can make sure it always exists.

File: c++/src/Reader.cc [hunk 11]
```diff
@@ -1205,8 +1215,32 @@ namespace orc {
...
+      auto fileSchema = &getSelectedType();
+      auto readType = schemaEvolution.getReadType();
+      std::set<uint64_t> readColumns;
+      getReadColumns(readType, readColumns);
```
Comment (wgtmac) [inline c++/src/Reader.cc:1222]: What about this?
```cpp
getReadColumns(readType, readColumnIds);
getReadColumns(getSelectedType(), selectedColumnIds);
if (readColumnIds != selectedColumnIds) {
  ...
}
```

File: c++/include/orc/Vector.hh (hunk 3 above)
Comment (ffacs) [reply c++/include/orc/Vector.hh:58]: But, what if the ColumnVectorBatch is created by another reader which was createed without the useTightNumericVector option?

Comment (wgtmac) [reply c++/include/orc/Vector.hh:58]: Does the current `ColumnReader` have same issue?

Comment (ffacs) [reply c++/include/orc/Vector.hh:58]: > Does the current `ColumnReader` have same issue?

Yes, I think there is.

Comment (wgtmac) [reply c++/include/orc/Vector.hh:58]: What probably need to fix that as well.

What about adding a new function to `ColumnVectorBatch` like `int32_t fixed_type_size() const` to return the width in bytes for fixed length type? For variable length type or nested type we can return -1.

File: c++/src/ConvertColumnReader.cc (hunk 7 above)
Comment (ffacs) [reply c++/src/ConvertColumnReader.cc:46]: > I guess this is not required

https://github.com/apache/orc/blob/070c1a5dce3fd7f6b6b8938cc3d872a7b67b1b28/c%2B%2B/src/MemoryPool.cc#L120-L127
When there are no nulls and we are going to set null when an overflow in conversion occurs, the `notNull` of other elements should be `1`, but they are `0` after resizing, so the `memset` is needed.

File: c++/include/orc/Vector.hh (hunk 3 above)
Comment (ffacs) [reply c++/include/orc/Vector.hh:58]: > What probably need to fix that as well.
> 
> What about adding a new function to `ColumnVectorBatch` like `int32_t fixed_type_size() const` to return the width in bytes for fixed length type? For variable length type or nested type we can return -1.

I think what we are going to do is handle incorrect usage by users (such as passing a vector that was created by other readers). However, we don’t know what causes the failure of casting. The vector may have been created by another reader with a different read type from the current reader, or it may have been created by another reader with a different `useTightNumericVector` option. `fixed_type_size()` may not help us infer what the user did wrong, but with `isTight`, we could suggest that the user check their options.

Comment (wgtmac) [reply c++/include/orc/Vector.hh:58]: OK, that sounds good. 

However, `isTight` here is a little bit confusing since it applies to only numeric types. We need to be careful to add any field to the public api. And it should also be used with its type together. 

In the `ColumnReader`, we always check the result of dynamic_cast of the vector batch. Is it enough if we can use dynamic_cast and check if it is not null in the `ConvertColumnReader`?

Comment (ffacs) [reply c++/include/orc/Vector.hh:58]: > OK, that sounds good.
> 
> However, `isTight` here is a little bit confusing since it applies to only numeric types. We need to be careful to add any field to the public api. And it should also be used with its type together.
> 
> In the `ColumnReader`, we always check the result of dynamic_cast of the vector batch. Is it enough if we can use dynamic_cast and check if it is not null in the `ConvertColumnReader`?

I think that`s OK, so we will only verify the `ReaderOption` and check the result of dynamic_cast in `ConvertColumnReader`.

File: c++/src/ColumnReader.cc [hunk 12]
```diff
@@ -828,7 +830,8 @@ namespace orc {
     std::vector<std::unique_ptr<ColumnReader>> children;
 
    public:
-    StructColumnReader(const Type& type, StripeStreams& stipe, bool useTightNumericVector = false);
+    StructColumnReader(const Type& type, StripeStreams& stipe, bool useTightNumericVector = false,
```
Comment (wgtmac) [inline c++/src/ColumnReader.cc:833]: I am thinking if we can use a new ColumnReaderOptions to include `useTightNumericVector` and `throwOnSchemaEvolutionOverflow` in case we will add more options. But that can be a separate patch since this is not public.

File: c++/src/Reader.cc [hunk 13]
```diff
@@ -1204,9 +1207,33 @@ namespace orc {
...
+    }
+  }
+
   std::unique_ptr<ColumnVectorBatch> RowReaderImpl::createRowBatch(uint64_t capacity) const {
-    return getSelectedType().createRowBatch(capacity, *contents->pool, enableEncodedBlock,
-                                            useTightNumericVector);
+    // If the read type is specified, then check that the selected schema matches the read type
+    // on the first call to createRowBatch.
+    if (schemaEvolution.getReadType() && selectedSchema.get() == nullptr) {
+      auto fileSchema = &getSelectedType();
```
Comment (wgtmac) [inline c++/src/Reader.cc:1221]: Move it to above line 1220.

File: c++/src/Reader.cc [hunk 14]
```diff
@@ -1204,9 +1207,33 @@ namespace orc {
...
+      getColumnIds(type->getSubtype(i), columnIds);
+    }
+  }
+
   std::unique_ptr<ColumnVectorBatch> RowReaderImpl::createRowBatch(uint64_t capacity) const {
-    return getSelectedType().createRowBatch(capacity, *contents->pool, enableEncodedBlock,
-                                            useTightNumericVector);
+    // If the read type is specified, then check that the selected schema matches the read type
+    // on the first call to createRowBatch.
+    if (schemaEvolution.getReadType() && selectedSchema.get() == nullptr) {
```
Comment (wgtmac) [inline c++/src/Reader.cc:1220]:
```suggestion
    auto selectedFileType = &getSelectedType();
    if (schemaEvolution.getReadType()) {
```

File: c++/src/Reader.cc [hunk 15]
```diff
@@ -1204,9 +1207,33 @@ namespace orc {
...
+      getColumnIds(fileSchema, fileColumns);
+      if (readColumns != fileColumns) {
+        std::ostringstream ss;
+        ss << "The selected schema " << fileSchema->toString() << " doesn't match read type "
+           << readType->toString();
+        throw SchemaEvolutionError(ss.str());
+      }
+    }
+    const Type& readType =
+        schemaEvolution.getReadType() ? *schemaEvolution.getReadType() : getSelectedType();
```
Comment (wgtmac) [inline c++/src/Reader.cc:1234]:
```suggestion
        schemaEvolution.getReadType() ? *schemaEvolution.getReadType() : *selectedFileType;
```

File: c++/src/Reader.cc [hunk 16]
```diff
@@ -1204,9 +1207,33 @@ namespace orc {
...
-    return getSelectedType().createRowBatch(capacity, *contents->pool, enableEncodedBlock,
-                                            useTightNumericVector);
+    // If the read type is specified, then check that the selected schema matches the read type
+    // on the first call to createRowBatch.
+    if (schemaEvolution.getReadType() && selectedSchema.get() == nullptr) {
+      auto fileSchema = &getSelectedType();
+      auto readType = schemaEvolution.getReadType();
+      std::set<uint64_t> readColumns, fileColumns;
+      getColumnIds(readType, readColumns);
+      getColumnIds(fileSchema, fileColumns);
```
Comment (wgtmac) [inline c++/src/Reader.cc:1225]:
```suggestion
      getColumnIds(selectedFileType, fileColumns);
```

File: c++/src/ConvertColumnReader.cc [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+  template <typename FileTypeBatch, typename ReadTypeBatch, typename ReadType>
+  class NumericConvertColumnReader : public ConvertColumnReader {
+   public:
+    NumericConvertColumnReader(const Type& _readType, const Type& fileType, StripeStreams& stripe,
+                               bool _throwOnOverflow)
+        : ConvertColumnReader(_readType, fileType, stripe, _throwOnOverflow) {}
+
+    void next(ColumnVectorBatch& rowBatch, uint64_t numValues, char* notNull) override {
+      ConvertColumnReader::next(rowBatch, numValues, notNull);
+      const auto& srcBatch = dynamic_cast<const FileTypeBatch&>(*data);
```
Comment (wgtmac) [inline c++/src/ConvertColumnReader.cc:130]: Can we wrap `dynamic_cast` to throw a clear error message if failed? Something like
```cpp
template <typename TO, typename FROM>
TO* safe_cast(FROM* from) {
  auto to = dynamic_cast<TO*>(from);
  if (to == nullptr) {
     throw ... // maybe a new exception type for cast in Exceptions.hh
  }
  return to;
}
```

**Review by wgtmac — COMMENTED** [review]
Rest LGTM.

File: c++/src/Reader.cc (hunk 13 above)
Comment (wgtmac) [reply c++/src/Reader.cc:1221]: OK, that sounds good. Sorry for not reading carefully.

File: c++/src/ConvertColumnReader.cc [hunk 18]
```diff
@@ -0,0 +1,445 @@
...
```
Comment (wgtmac) [inline c++/src/ConvertColumnReader.cc:143]: It may also throw here, better to apply `SafeCastBatchTo` as well.

File: c++/src/ConvertColumnReader.cc [hunk 19]
```diff
@@ -0,0 +1,445 @@
...
+  class NumericConvertColumnReader<FileTypeBatch, BooleanVectorBatch, bool>
+      : public ConvertColumnReader {
+   public:
+    NumericConvertColumnReader(const Type& _readType, const Type& fileType, StripeStreams& stripe,
+                               bool _throwOnOverflow)
+        : ConvertColumnReader(_readType, fileType, stripe, _throwOnOverflow) {}
+
+    void next(ColumnVectorBatch& rowBatch, uint64_t numValues, char* notNull) override {
+      ConvertColumnReader::next(rowBatch, numValues, notNull);
+      const auto& srcBatch = dynamic_cast<const FileTypeBatch&>(*data);
```
Comment (wgtmac) [inline c++/src/ConvertColumnReader.cc:172]: ditto

File: c++/src/ConvertColumnReader.cc [hunk 20]
```diff
@@ -0,0 +1,445 @@
...
+    if (static_cast<int64_t>(dstValue) != inputLong) {
+      return false;
+    }
+    return true;
+  }
+
+  template <typename DestBatchRefType>
+  static inline DestBatchRefType SafeCastBatchTo(ColumnVectorBatch& batch, const Type& readType) {
+    try {
+      return dynamic_cast<DestBatchRefType>(batch);
```
Comment (wgtmac) [inline c++/src/ConvertColumnReader.cc:91]: Why not casting to pointer type and checks null directly? If would be more efficient to avoid try-catch on every cast.

Comment (ffacs) [reply c++/src/ConvertColumnReader.cc]: Done

**Review by wgtmac — APPROVED** [review]
LGTM. Thanks!

File: c++/src/Reader.hh [hunk 21]
```diff
@@ -237,6 +241,10 @@ namespace orc {
     bool getThrowOnHive11DecimalOverflow() const;
     bool getIsDecimalAsLong() const;
     int32_t getForcedScaleOnHive11Decimal() const;
+
+    const SchemaEvolution& getSchemaEvolution() const {
```
Comment (coderex2522) [inline c++/src/Reader.hh:245]: Is it possible to consider unifying the return type of the getSchemaEvolution() function in class StripeStreams and class RowReaderImpl.

Comment (ffacs) [reply c++/src/Reader.hh]: Done

File: c++/src/SchemaEvolution.cc [hunk 22]
```diff
@@ -16,6 +16,8 @@
  * limitations under the License.
  */
 
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (ffacs) [reply c++/src/SchemaEvolution.cc]: Removed

**Review by coderex2522 — APPROVED** [review]
+1 LGTM. Thanks @ffacs

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, All.

This code seems to cause a build failure on Clang-Ubuntu environment. Not only old Clang 10.0.0 on Ubuntu 20.04 but also Clang 14.0.0 on Ubuntu 22.04 complains about this patch. Could you take a look at this?

**Comment by dongjoon-hyun** [comment]
Could you double check this compilation issue, @ffacs , @wgtmac , @coderex2522 ?

File: c++/src/ConvertColumnReader.cc [hunk 23]
```diff
@@ -0,0 +1,446 @@
...
+      throw SchemaEvolutionError(
+          "SchemaEvolution only support tight vector, please create ColumnVectorBatch with "
+          "option useTightNumericVector");
+    }
+    const auto& _readType = *stripe.getSchemaEvolution()->getReadType(fileType);
+
+    switch (fileType.getKind()) {
+      case BOOLEAN: {
+        switch (_readType.getKind()) {
+          CASE_CREATE_READER(BYTE, BooleanToByte);
```
Comment (dongjoon-hyun) [inline c++/src/ConvertColumnReader.cc:260]: For example, `;` is required in this line?

Comment (dongjoon-hyun) [reply c++/src/ConvertColumnReader.cc:260]: This is fixed via https://github.com/apache/orc/pull/1500#discussion_r1238261627

Comment (wgtmac) [reply c++/src/ConvertColumnReader.cc:260]: How did you get the failure? Via docker test?
